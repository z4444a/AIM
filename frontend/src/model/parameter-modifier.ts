export enum ParameterModifier {
  POOL_PARAMETER = 1,
  REQUEST_PARAMETER = 2,
  ALLOCATION_PARAMETER = 3,
}
