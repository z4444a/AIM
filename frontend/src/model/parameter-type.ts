export enum ParameterType {
  NUMBER = 'NUMBER',
  TEXT = 'TEXT',
  DATE = 'DATE',
  LIST = 'LIST',
  REAL = 'REAL',
  POOL = 'POOL',
}
