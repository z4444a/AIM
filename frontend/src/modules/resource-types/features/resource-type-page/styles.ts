import createStyles from '@material-ui/core/styles/createStyles';
import { Theme, WithStyles } from '@material-ui/core';

export const styles = (theme: Theme) =>
  createStyles({
    container: {
      display: 'flex',
      justifyContent: 'center',
      '& > div': {
        width: '95%',
        minWidth: theme.breakpoints.values.sm,
      },
    },
  });

export type Style = WithStyles<'container'>;
