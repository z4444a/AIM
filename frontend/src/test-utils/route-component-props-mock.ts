import { RouteComponentProps } from 'react-router';
import { Href } from 'history';

export function getMockRouterProps<P>(data: P): RouteComponentProps {
  const location = {
    hash: '',
    key: '',
    pathname: '',
    search: '',
    state: {},
  };

  const props: RouteComponentProps<P> = {
    match: {
      isExact: true,
      params: data,
      path: '',
      url: '',
    },
    location: location,
    history: {
      length: 2,
      action: 'POP',
      location: location,
      push: () => {},
      replace: () => {},
      go: () => {},
      goBack: () => {},
      goForward: () => {},
      block: () => () => {},
      createHref: () => {
        const temp: Href = '';
        return temp;
      },
      listen: () => () => {},
    },
    staticContext: {},
  };

  return props;
}
