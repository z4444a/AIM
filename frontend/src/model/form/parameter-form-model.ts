import { ParameterConstraintFormModel } from './parameter-constraint-form-model';
import { ParameterType } from '../parameter-type';
import { ParameterModifier } from '../parameter-modifier';

export interface ParameterFormModel {
  id?: number;
  key?: number;
  name: string;
  identifier: string;
  required: boolean;
  visibleToOwner?: boolean;
  modifier: ParameterModifier;
  parameterType: ParameterType;
  constraint: ParameterConstraintFormModel;
  order: number;
  poolTypeId?: number;
}
