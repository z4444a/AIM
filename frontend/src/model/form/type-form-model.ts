import { ParameterFormModel } from './parameter-form-model';
import { PictureFormModel } from './picture-form-model';

export interface TypeFormModel {
  id?: number;
  name: string;
  description: string;
  active: boolean;
  needsBackup: boolean;
  quantitative: boolean;
  picture: PictureFormModel;
  parameters: ParameterFormModel[];
}
