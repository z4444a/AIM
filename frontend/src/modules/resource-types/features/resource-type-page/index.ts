import withStyles from '@material-ui/core/styles/withStyles';
import { styles } from './styles';
import { ResourceTypePage } from './resource-type-page';
import { push } from 'connected-react-router';
import { connect } from 'react-redux';
import {
  addParameterRow,
  changeParametersOrder,
  findParameterIdentifier,
  hideDialog,
  removeParameterRow,
  resetCheckUniqueTypeName,
  resetEditedParameterName,
  resetFormModel,
  resetHighlightEmpty,
  resetParameterFormModel,
  setEditable,
  setEditedParameterName,
  setHighlightEmpty,
  showDialog,
  updateFormModel,
  updateParameterFormModel,
} from '../../../../redux/actions/resource-type-page-actions';
import { RootState } from 'typesafe-actions';
import { ParameterFormModel } from '../../../../model/form/parameter-form-model';
import { TypeFormModel } from '../../../../model/form/type-form-model';
import { ParameterModifier } from '../../../../model/parameter-modifier';
import { checkTypeName } from '../../../../redux/actions/exists-type-name-action';
import { ParameterRowFormModel } from '../../../../model/form/parameter-row-form-model';
import { hidePortalOverlay, showPortalOverlay } from '../../../../redux/actions/portal-actions';
import { fetchResourceTypesSuggestions } from '../../../../redux/actions/resource-pool-create-page-actions';
import { ParameterUniqueFields } from '../../../../model/parameters/parameter-unique-fields';
import { ValidationParameterResponse } from '../../../../model/validation-parameter-response';

export const mapDispatchToProps = (dispatch: Function) => ({
  updateParameterFormModel: (index: number, model: ParameterFormModel) =>
    dispatch(updateParameterFormModel({ key: index, model })),
  removeParameterFormModel: (index: number) => dispatch(removeParameterRow(index)),
  resetParameterFormModel: () => dispatch(resetParameterFormModel()),
  updateFormModel: (type: TypeFormModel) => dispatch(updateFormModel(type)),
  resetFormModel: () => dispatch(resetFormModel()),
  setEditedParameterName: (name: string) => dispatch(setEditedParameterName(name)),
  resetEditedParameterName: () => dispatch(resetEditedParameterName()),
  checkTypeName: (name: string) => dispatch(checkTypeName.request(name)),
  resetCheckUniqueTypeName: () => dispatch(resetCheckUniqueTypeName()),
  setRowEditable: (key: number, editable: boolean) => dispatch(setEditable({ key, editable })),
  addParameterRow: (modifier: ParameterModifier) => dispatch(addParameterRow(modifier)),
  showDialog: () => dispatch(showDialog()),
  hideDialog: () => dispatch(hideDialog()),
  setHighlightEmpty: () => dispatch(setHighlightEmpty()),
  resetHighlightEmpty: () => dispatch(resetHighlightEmpty()),
  redirect: (path: string) => dispatch(push(path)),
  changeOrder: (rows: ParameterRowFormModel[]) => dispatch(changeParametersOrder(rows)),

  onShowOverlay: (text: string) => dispatch(showPortalOverlay(text)),
  onHideOverlay: () => dispatch(hidePortalOverlay()),

  fetchResourceTypesSuggestions: () => dispatch(fetchResourceTypesSuggestions.request()),

  findParamIdentifier: (params: ParameterUniqueFields) =>
    dispatch(findParameterIdentifier.request(params)),
  setValidation: (params: ValidationParameterResponse | null) =>
    dispatch(findParameterIdentifier.success(params)),
});

export const mapStateToProps = (state: RootState) => ({
  typeSuggestions: state.resourcePoolCreatePage.resourceTypeSuggestion.suggestions,
  formModel: state.resourceTypePage.formModel,
  uniqueName: state.resourceTypePage.uniqueName,
  editedParameterName: state.resourceTypePage.editedParameterName,
  parameterRows: state.resourceTypePage.parameterRows,
  dialogIsOpen: state.resourceTypePage.dialogIsOpen,
  highlightEmpty: state.resourceTypePage.highlightEmpty,
  removedParameterResponse: state.resourceTypePage.removedParameterResponse,
  identifierValidation: state.resourceTypePage.identifierValidation,
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withStyles(styles)(ResourceTypePage));
