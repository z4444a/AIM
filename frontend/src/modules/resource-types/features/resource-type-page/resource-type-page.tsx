import React from 'react';
import { Style } from './styles';
import ResourceTypesCard from '../../components/resource-type-card/index';
import { mapDispatchToProps, mapStateToProps } from './index';
import { Path } from '../../../../commons/path';
import { TypeFormModel } from '../../../../model/form/type-form-model';
import { Props as ResourceTypeCardProps } from '../../components/resource-type-card/card';
import _ from 'underscore';

export interface Props {
  onSubmit: (type: TypeFormModel) => void;
  showIdentifiers?: boolean;
  oldModel?: TypeFormModel;

  findParamValue?: (id: number) => void;
  resetRemoveParameterResponse?: () => void;
}

export type InternalProps = Props &
  ReturnType<typeof mapDispatchToProps> &
  ReturnType<typeof mapStateToProps> &
  Style;

export class ResourceTypePage extends React.PureComponent<InternalProps> {
  public componentDidMount(): void {
    const { fetchResourceTypesSuggestions } = this.props;
    if (typeof fetchResourceTypesSuggestions === 'function') {
      fetchResourceTypesSuggestions();
    }
  }

  public componentWillUnmount(): void {
    const {
      resetEditedParameterName,
      resetFormModel,
      resetHighlightEmpty,
      resetParameterFormModel,
      resetCheckUniqueTypeName,
    } = this.props;
    resetParameterFormModel();
    resetFormModel();
    resetEditedParameterName();
    resetCheckUniqueTypeName();
    resetHighlightEmpty();
  }

  public render(): React.ReactNode {
    const {
      classes,
      parameterRows,
      formModel,
      uniqueName,
      editedParameterName,
      highlightEmpty,
      showIdentifiers,
      updateFormModel,
      setRowEditable,
      removeParameterFormModel,
      updateParameterFormModel,
      addParameterRow,
      setEditedParameterName,
      resetEditedParameterName,
      showDialog,
      hideDialog,
      setHighlightEmpty,
      resetHighlightEmpty,
      checkTypeName,
      dialogIsOpen,
      changeOrder,

      onShowOverlay,
      onHideOverlay,

      removedParameterResponse,
      findParamValue,
      resetRemoveParameterResponse,
      typeSuggestions,

      findParamIdentifier,
      identifierValidation,
      setValidation,
    } = this.props;
    const cardProps: ResourceTypeCardProps = {
      dialogIsOpen: dialogIsOpen,
      editedParameterName: editedParameterName,
      highlightEmpty: highlightEmpty,
      onCancelButtonClick: this.handleRedirect,
      onAddParameterRow: addParameterRow,
      onChange: updateFormModel,
      onNameValidationRequested: checkTypeName,
      onCompleteButtonClick: this.handleSubmit,
      onDialogClosed: hideDialog,
      onDialogOpened: showDialog,
      onHighlightEmptyReset: resetHighlightEmpty,
      onHighlightEmptySet: setHighlightEmpty,
      onParameterEditReset: resetEditedParameterName,
      onParameterEditSet: setEditedParameterName,
      onParameterFormChange: updateParameterFormModel,
      onParameterRowEditSet: setRowEditable,
      onParameterRowRemoval: removeParameterFormModel,
      parameterRows: parameterRows,
      showIdentifiers: showIdentifiers,
      type: formModel,
      uniqueName: uniqueName,
      changeOrder: changeOrder,

      onShowOverlay: onShowOverlay,
      onHideOverlay: onHideOverlay,

      removedParameterResponse: removedParameterResponse,
      findParamValue: findParamValue,
      resetRemoveParameterResponse: resetRemoveParameterResponse,
      typeSuggestions: typeSuggestions,

      findParamIdentifier: findParamIdentifier,
      identifierValidation: identifierValidation,
      setValidation: setValidation,
    };
    return (
      <div className={classes.container}>
        <ResourceTypesCard {...cardProps} />
      </div>
    );
  }

  private handleSubmit = () => {
    const { onSubmit, formModel, oldModel, parameterRows } = this.props;
    if (_.isEqual(oldModel, formModel)) {
      this.handleRedirect();
    }
    const { parameters, ...rest } = formModel;
    const formToSubmit: TypeFormModel = {
      parameters: parameterRows.map(value => value.model),
      ...rest,
    };
    onSubmit(formToSubmit);
  };

  private handleRedirect = () => {
    const { redirect, resetHighlightEmpty, resetFormModel } = this.props;
    redirect(Path.RESOURCE_TYPES);
    resetFormModel();
    resetHighlightEmpty();
  };
}
