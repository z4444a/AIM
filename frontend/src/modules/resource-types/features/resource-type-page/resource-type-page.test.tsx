import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { InternalProps, ResourceTypePage } from './resource-type-page';
import { getMockRouterProps } from '../../../../test-utils/route-component-props-mock';
import { TypeFormModel } from '../../../../model/form/type-form-model';
import { ParameterFormModel } from '../../../../model/form/parameter-form-model';
import { ParameterType } from '../../../../model/parameter-type';
import { ParameterModifier } from '../../../../model/parameter-modifier';

describe('ResourceTypePage', () => {
  let wrapper: ShallowWrapper;
  beforeEach(() => {
    const formModel: TypeFormModel = {
      active: false,
      description: '',
      id: 0,
      name: '',
      needsBackup: false,
      quantitative: false,
      parameters: [],
      picture: {
        picture: [],
        picturePath: '',
      },
    };

    const parameterFormModel: ParameterFormModel = {
      constraint: {},
      identifier: '',
      name: '',
      parameterType: ParameterType.NUMBER,
      required: false,
      modifier: ParameterModifier.REQUEST_PARAMETER,
      order: 0,
    };

    const props: InternalProps = {
      ...getMockRouterProps(null),
      classes: {
        container: 'container',
      },
      dialogIsOpen: false,
      editedParameterName: null,
      formModel: formModel,
      highlightEmpty: false,
      parameterFormModel: parameterFormModel,
      hideDialog: () => null,
      onSubmit: (p1: TypeFormModel) => null,
      redirect: () => null,
      resetEditedParameterName: () => null,
      resetFormModel: () => null,
      resetHighlightEmpty: () => null,
      resetParameterFormModel: () => null,
      setEditedParameterName: () => null,
      setHighlightEmpty: () => null,
      showDialog: () => null,
      updateFormModel: () => null,
      updateParameterFormModel: () => null,
    };
    wrapper = shallow(<ResourceTypePage {...props} />);
  });

  it('should render correctly', () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
