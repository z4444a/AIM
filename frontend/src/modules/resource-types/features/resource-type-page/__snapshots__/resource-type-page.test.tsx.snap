// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResourceTypePage should render correctly 1`] = `
<div
  className="container"
>
  <withI18nextTranslation(WithStyles(ResourceTypesCard))
    dialogIsOpen={false}
    editedParameterName={null}
    highlightEmpty={false}
    onCancelButtonClick={[Function]}
    onChange={[Function]}
    onCompleteButtonClick={[Function]}
    onDialogClosed={[Function]}
    onDialogOpened={[Function]}
    onHighlightEmptyReset={[Function]}
    onHighlightEmptySet={[Function]}
    onParameterEditReset={[Function]}
    onParameterEditSet={[Function]}
    onParameterFormChange={[Function]}
    type={
      Object {
        "active": false,
        "description": "",
        "id": 0,
        "name": "",
        "needsBackup": false,
        "parameters": Array [],
        "picture": Object {
          "picture": Array [],
          "picturePath": "",
        },
        "quantitative": false,
      }
    }
  />
</div>
`;
